README.rst
django-jackfrost 0.4.0
======================

Convert your Django views into a collection of static HTML files. Or, put
another way, a Django based static-site-generator with few opinions.

This is my attempt at a different type of static site renderer, instead
leveraging the availability of Django's `staticfiles`_ functionality to leave
specifics to someone else :)
The theory is thus that you could choose a third party storage from, say,
`django-storages`_ and plug it into ``jackfrost`` and have things Just Work.

I don't actually know if that's true though.

.. |travis_stable| image:: https://travis-ci.org/kezabelle/django-jackfrost.svg?branch=0.4.0
  :target: https://travis-ci.org/kezabelle/django-jackfrost

.. |travis_master| image:: https://travis-ci.org/kezabelle/django-jackfrost.svg?branch=master
  :target: https://travis-ci.org/kezabelle/django-jackfrost


==============  ======
Release         Status
==============  ======
stable (0.4.0)  |travis_stable|
master          |travis_master|
==============  ======

Alternatives
============

- There's `django-medusa`_, though it doesn't appear to be active anymore.
- also `django-bakery`_, which takes a different approach whereby one must
  extend specific views or models.

Differences
-----------

- Unlike `django-medusa`_ there is no autodiscovery, and no requirement that
  renderers go in a specific place.
- Unlike `django-bakery`_, existing views and models ought to be usable unchanged,
  because of the renderer approach I've taken, which is more similar to
  `django-medusa`_ or `Django RSS Feeds`_.

Dependencies
============

-  `Django`_
-  `pytest`_

Installing
==========

`pip`_ installing the latest release via PyPI::

    pip install django-jackfrost==0.4.0


If you want to get the latest, unstable version, you can use
something like this (again with `pip`_) I think::

    pip install git+https://github.com/kezabelle/django-jackfrost.git#egg=django-jackfrost


Put ``jackfrost`` into your ``INSTALLED_APPS``::

    INSTALLED_APPS = (
        'django.contrib.auth',
        # ...
        'jackfrost',
        # ...
    )

which will enable the management command::

    python manage.py collectstaticsite --processes=N


Configuration & usage
---------------------

Set ``JACKFROST_STORAGE`` to whatever storage backend you'd like to use, in
your project's settings. By default, a subclass of
``django.contrib.staticfiles.storage.StaticFilesStorage`` which puts output into
a ``__jackfrost`` directory will be used.

If your storage backend needs any arguments that can't be gleaned from individual
settings, you can set ``JACKFROST_STORAGE_KWARGS`` to a dictionary of
arguments to be used when instantiating the ``JACKFROST_STORAGE``


Selecting renderers
^^^^^^^^^^^^^^^^^^^

Add a ``JACKFROST_RENDERERS`` setting, which should be a list or tuple of
dotted paths to python classes or functions, much like ``MIDDLEWARE_CLASSES``,
``TEMPLATE_CONTEXT_PROCESSORS`` etc::

    JACKFROST_RENDERERS = (
        'myapp.renderers.MyModelRenderer',
        'my_other_app.utils.SomeOtherRenderer',
    )

In theory, I don't care whether your ``JACKFROST_RENDERERS`` are functions
or classes; if it's a class it must implement ``__call__``. Either way,
it should, when called, return a number of URL paths to be consumed.


Renderers for models
^^^^^^^^^^^^^^^^^^^^

If you have a model which has a ``get_absolute_url`` method, your renderer
can be as simple as::

    from jackfrost.models import ModelRenderer

    class MyModelRenderer(ModelRenderer):
        def get_model(self):
            return MyModel

If you need to customise the queryset, there is a ``get_queryset`` method
which can be replaced. There is also a ``get_urls`` method, if you need to
go totally off-reservation.


Reading from `sitemaps`_
^^^^^^^^^^^^^^^^^^^^^^^^

Giving ``jackfrost`` the dotted path to a standard `Django sitemap`_ as
one of the ``JACKFROST_RENDERERS`` should do the right thing, and get the
URLs out of the sitemap itself without you needing to do anything or write
a new renderer.


Reading from `django-medusa`_
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In theory, giving ``jackfrost`` the dotted path to a subclass of the `django-medusa`_
`BaseStaticSiteRenderer`_ should do the right thing, and get the URLs out of
the medusa renderer itself, without you doing anything. It will avoid going
through the medusa rendering process, instead it'll go through mine.


Reading from `Django RSS Feeds`_
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Giving ``jackfrost`` the dotted path to a subclass of a `Feed`_
should do the right thing, and get the URLs out by asking the `Feed`_ for the
``item_link`` for everything in ``items``, without you doing anything.


Writing a renderer
^^^^^^^^^^^^^^^^^^

The most basic renderer would be::

    def myrenderer_yielding():
        yield reverse('app:name')

or::

    def myrenderer():
        return [reverse('app:name')]

Renderers may also be classes::

    class MyRenderer(object):
        __slots__ = ()

        def __init__(self):
            pass

        def __call__(self):
            yield reverse('app:name')


Listening for renders
^^^^^^^^^^^^^^^^^^^^^

There are 8 signals in total:

* ``build_started`` is fired when the management command is run.
* ``reader_started`` is fired when a ``URLReader`` instance begins working.
* ``read_page`` is fired when a ``URLReader`` successfully gets a URL's content.
* ``reader_finished`` is fired when a ``URLReader`` instance completes.
* ``writer_started`` is fired when a ``URLWriter`` instance begins working.
* ``write_page`` is fired just after the content is written to the storage backend.
* ``writer_finished`` is fired when the ``URLWriter`` completes
* ``build_finished`` fires at the end of the management command.

Rendering on model change
^^^^^^^^^^^^^^^^^^^^^^^^^

Additionally, there is a listener, ``jackfrost.receivers.build_page_for_obj`` which
is suitable for being used as a ``pre_save`` or ``post_save`` receiver for
a ``Model`` instance, and will attempt to build just the ``get_absolute_url`` for
that object.

Defining when a model may build
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If a ``Model`` instance implements a ``jackfrost_can_build`` method, this is
checked before building the static page. If ``jackfrost_can_build`` returns
``False``, the page won't get built. Any other value will result in it being
built.

Defining different URLs
^^^^^^^^^^^^^^^^^^^^^^^

If a ``Model`` instance implements a ``jackfrost_urls`` method, this
is used instead of the ``get_absolute_url``, and should return an iterable of
all the URLs to consider building.

If the ``Model`` instance has a ``get_list_url`` method, that page will also be
built. Useful for updating any ``ListView`` pages, etc.

Extras
------

Where possible, ``jackfrost`` will attempt to compensate for redirects (301, 302 etc)
by writing an HTML page with a ``<meta refresh>`` tag pointing at the final
endpoint. The template used is called `301.html`.

Additionally, static pages for 401, 403, 404 and 500 errors will be built
from their respective templates, if they exist. Useful if you want to wire
up Apache ``ErrorDocument`` directives or whatever.


Running the tests (87% coverage)
--------------------------------

Given a complete clone::

    python setup.py test

.. _django-medusa: https://github.com/mtigas/django-medusa
.. _staticfiles: https://docs.djangoproject.com/en/stable/ref/contrib/staticfiles/
.. _Django: https://docs.djangoproject.com/en/stable/
.. _pip: https://pip.pypa.io/en/stable/
.. _django-storages: https://django-storages.readthedocs.org/en/latest/
.. _pytest: http://pytest.org/latest/
.. _django-bakery: http://django-bakery.readthedocs.org/en/latest/
.. _they don't: https://github.com/datadesk/django-bakery/issues/15
.. _sitemaps: https://docs.djangoproject.com/en/stable/ref/contrib/sitemaps/
.. _Django sitemap: https://docs.djangoproject.com/en/stable/ref/contrib/sitemaps/
.. _BaseStaticSiteRenderer: https://github.com/mtigas/django-medusa/blob/master/django_medusa/renderers/base.py
.. _Django RSS Feeds: https://docs.djangoproject.com/en/stable/ref/contrib/syndication/
.. _Feed: https://docs.djangoproject.com/en/stable/ref/contrib/syndication/#feed-class-reference


CHANGELOG
Change log
----------

?next?
^^^^^^

* Added a ``--dry-run`` argument to the ``collectstaticsite`` management command
  to preview changes using the standard Django server, without writing anything
  to the storage backend.
* Moved ``jackfrost.utils.build_page_for_obj`` and ``jackfrost.utils.eventlog_write`
  to ``receivers`` to better represent their use as signal listeners.
* Raise an error if a URL doesn't end with a slash or a file extension.

0.4.0
^^^^^^

* Added ``signals.write_page`` which is fired after the storage backend has
  written the content.
* Added ``utils.eventlog_write`` for using `pinax.eventlog`_ and the
  aforementioned ``write_page`` signal to track built items.
* Added a `ModelAdmin action`_ for building a selection of model instances
  into their static counterparts, from the admin changelist.
* Made the receiver ``utils.build_page_for_obj`` actually work.
* the ``models`` module will be compiled using Cython, if installed.

.. _pinax.eventlog: https://github.com/pinax/pinax-eventlog
.. _ModelAdmin action: https://docs.djangoproject.com/en/stable/ref/contrib/admin/actions/

0.3.0
^^^^^^

* Replaced ``jackfrost_absolute_url`` method in ``ModelRenderer`` with
  ``jackfrost_urls``, which allows models to add every URL they want to cause
  to be visited.


0.2.1
^^^^^^

* Added support for Django 1.6
* Added celery tasks for building both a single URL, and all URLs.

0.2.0
^^^^^^

* Initial release.



jackfrost/models.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import
from __future__ import print_function
from __future__ import unicode_literals
from __future__ import division
from collections import namedtuple
import hashlib
import logging
from mimetypes import guess_extension
# from django.core.urlresolvers import set_script_prefix
from django.conf.urls import url
from django.core.exceptions import ImproperlyConfigured
from django.core.files.base import ContentFile
from django.http import HttpResponse, StreamingHttpResponse
from django.template import TemplateDoesNotExist
from django.template.loader import render_to_string
from django.utils.encoding import force_bytes
from django.utils.encoding import force_text
from django.utils.http import is_safe_url
# noinspection PyUnresolvedReferences
from django.utils.six.moves.urllib.parse import urlparse
from jackfrost.signals import reader_started
from jackfrost.signals import read_page
from jackfrost.signals import reader_finished
from jackfrost.signals import writer_started
from jackfrost.signals import write_page
from jackfrost.signals import writer_finished
from jackfrost.utils import is_url_usable
from os.path import splitext
try:
    from django.utils.module_loading import import_string
except ImportError:  # pragma: no cover
    from django.utils.module_loading import import_by_path as import_string
from django.conf import settings
from django.core.paginator import Paginator
from django.test.client import Client
from django.utils import six
from jackfrost import defaults
from posixpath import normpath


__all__ = ['URLCollector', 'URLReader', 'ErrorReader', 'URLWriter',
           'ModelRenderer', 'SitemapRenderer', 'MedusaRenderer', 'FeedRenderer']
logger = logging.getLogger(__name__)


class CollectionError(ValueError): pass
class ReaderError(ValueError): pass


class ReadResult(namedtuple('ReadResult', 'url filename status content')):
    __slots__ = ()

    def as_response(self, request):
        return StreamingHttpResponse(streaming_content=self.content,
                                     status=self.status)

    def as_urls(self):
        yield url(regex=r'^{url}$'.format(url=self.url[1:]),
                  name=None, view=self.as_response)
        yield url(regex=r'^{url}$'.format(url=self.filename),
                  name=None, view=self.as_response)


# this is a class rather than a namedtuple instance itself because otherwise
# multiprocessint cannot handle cythonized versions:
# Reason: 'PicklingError("Can't pickle <class 'importlib.WriteResult'>",)'
class WriteResult(namedtuple('WriteResult', 'name created modified md5 storage_result')):
    __slots__ = ()


class URLReader(object):
    """
    Given a list of URLs, presumably from a URLCollector, build them to files
    """
    __slots__ = ('urls', '_client', '_content_types')

    def __init__(self, urls):
        self.urls = tuple(urls)
        self._client = None
        self._content_types = None

    def __repr__(self):
        num = len(self.urls)
        top3 = self.urls[0:3]
        urls_themselves = ', '.join('"%s"' % x for x in top3)
        remaining = num - len(top3)
        urls = '[%(top3)s'
        if remaining > 0:
            urls += ' ... %(more)d remaining'
        urls += ']'
        return '<%(mod)s.%(cls)s urls=%(urls)s>' % {
            'mod': self.__module__,
            'cls': self.__class__.__name__,
            'urls': urls % {'top3': urls_themselves, 'more': remaining},
        }

    def get_target_filename(self, url, response):
        if not is_url_usable(url):
            raise ReaderError(
                "the URL '%(url)s' does not end in a forward-slash ('/'), nor "
                "does it have a file extension." % {'url': url})
        path_part, extension = splitext(url)
        if extension == '':
            content_type = response['Content-Type']
            major = content_type.split(';', 1)[0]
            extension = self.content_types.get(
                major, guess_extension(major, strict=False))
            filename = '{path}/{file}{ext}'.format(path=url, file='index',
                                                   ext=extension)
        else:
            filename = url
        filename = filename.lstrip('/')
        return normpath(filename)

    @property
    def client(self):
        if self._client is None:
            self._client = Client()
        return self._client

    @property
    def content_types(self):
        if self._content_types is None:
            self._content_types = getattr(settings, 'JACKFROST_CONTENT_TYPES',
                                          defaults.JACKFROST_CONTENT_TYPES)
        return self._content_types

    def build_redirect_page(self, url, final_url):
        urlparts = urlparse(url)
        url = urlparts.path

        if not is_safe_url(url):
            logger.error("Unable to generate a redirecting page for {url} "
                         "because Django says it's not safe".format(url=url))
            return None

        try:
            result = render_to_string([
                normpath('{}/301.html'.format(final_url)),
                normpath('{}/301.html'.format(url)),
                '301.html',
            ], {'this_url': url, 'next_url': final_url})
        except TemplateDoesNotExist:
            logger.error("Unable to generate a redirecting page for {url} "
                         "because there is no 301 template".format(url=url),
                         exc_info=1)
            return None

        response = HttpResponse(content=result, content_type='text/html')
        filename = self.get_target_filename(url=url, response=response)
        return ReadResult(url=url, filename=filename, status=None,
                          content=force_bytes(result))

    def build_page(self, url):
        resp = self.client.get(url, follow=True)
        assert resp.status_code == 200, "Got %(code)d response for %(url)s" % {
            'code': resp.status_code, 'url': url}

        # calculate changed URL and redirects if necessary
        if hasattr(resp, 'redirect_chain') and resp.redirect_chain:
            previous_pages = resp.redirect_chain[:]
            previous_pages.insert(0, (url, 301))  # hack to put in *this* url.
            final_url = previous_pages.pop()[0]
            urlparts = urlparse(final_url)
            url = urlparts.path
            for previous_page, previous_status in previous_pages:
                yield self.build_redirect_page(url=previous_page,
                                               final_url=url)

        filename = self.get_target_filename(url=url, response=resp)
        if resp.streaming is True:
            response_content = b''.join(resp.streaming_content)
        else:
            response_content = resp.content
        read_page.send(sender=self.__class__,
                       instance=self,
                       url=url,
                       response=resp,
                       filename=filename)
        yield ReadResult(url=url, filename=filename, status=resp.status_code,
                         content=force_bytes(response_content))

    def build(self):
        reader_started.send(sender=self.__class__, instance=self)
        for idx, url in enumerate(self.urls, start=0):
            for result in self.build_page(url=url):
                yield result
        reader_finished.send(sender=self.__class__, instance=self)

    def __call__(self):
        return self.build()


class ErrorReader(object):
    __slots__ = ('reader',)

    def __init__(self, reader_cls=None):
        if reader_cls is None:
            reader_cls = URLReader
        self.reader = reader_cls(urls=())

    def __call__(self):
        reader_started.send(sender=self.__class__, builder=self)
        for error in (401, 403, 404, 500):
            yield self.build_error_page(error=error)
        reader_finished.send(sender=self.__class__, builder=self)

    def build_error_page(self, error):
        try:
            result = render_to_string([
                '{error!s}.html'.format(error=error),
            ], {'request_path': None})
        except TemplateDoesNotExist:
            logger.error("Unable to generate a {error!s} page".format(error=error),  # noqa
                         exc_info=1)
            return None

        response = HttpResponse(content=result, status=error)
        filename = self.reader.get_target_filename(
            url='{error!s}.html'.format(error=error),
            response=response)
        read_page.send(sender=self.__class__,
                       instance=self,
                       url=None,
                       response=response,
                       filename=filename)
        return ReadResult(url=None, filename=filename, status=error,
                          content=result)



class URLWriter(object):
    __slots__ = ('data', '_storage')

    def __init__(self, data):
        self.data = data
        self._storage = None

    def __repr__(self):
        num = len(self.data)
        top3 = self.data[0:3]
        remaining = num - len(top3)
        urls_themselves = ', '.join('"%s"' % x.url for x in top3)
        urls = '(%(top3)s'
        if remaining > 0:
            urls += ' ... %(more)d remaining'
        urls += ')'
        return '<%(mod)s.%(cls)s data=%(urls)s>' % {
            'mod': self.__module__,
            'cls': self.__class__.__name__,
            'urls': urls % {'top3': urls_themselves, 'more': remaining},
        }


    @property
    def storage(self):
        if self._storage is None:
            storage = getattr(settings, 'JACKFROST_STORAGE',
                          defaults.JACKFROST_STORAGE)
            kwargs = getattr(settings, 'JACKFROST_STORAGE_KWARGS',
                             defaults.JACKFROST_STORAGE_KWARGS)
            storage_cls = import_string(storage)
            self._storage = storage_cls(**kwargs)
        return self._storage

    def write(self, data):
        """
        :type data: jackfrost.models.ReadResult
        """
        name = data.filename
        content = force_bytes(data.content)
        content_hash = hashlib.md5(content).hexdigest()
        content_io = ContentFile(content)

        file_exists = self.storage.exists(name=name)
        if file_exists:
            self.storage.delete(name=name)
        result = self.storage.save(name=name, content=content_io)
        write_result = WriteResult(name=name, created=not file_exists,
                                   modified=True, md5=content_hash,
                                   storage_result=result)
        write_page.send(sender=self.__class__, instance=self, read_result=data,
                        write_result=write_result)
        return write_result

    def build(self):
        writer_started.send(sender=self.__class__, instance=self)
        for idx, data in enumerate(self.data, start=0):
            write_result = self.write(data)
            yield write_result
        writer_finished.send(sender=self.__class__, instance=self)

    def __call__(self):
        return self.build()


class URLCollector(object):
    """
    >>> x = URLCollector(renderers=['a.b.C', DEF])
    >>> urls = x()
    ('/', '/a/b/', '/c/')
    """
    __slots__ = ('renderers',)

    def __init__(self, renderers=None):
        # TODO: figure out prefix necessities
        # url_prefix = getattr(settings, 'JACKFROST_SCRIPT_PREFIX', None)
        # if url_prefix is not None:
        #     set_script_prefix(url_prefix)
        self.renderers = frozenset(self.get_renderers(renderers=renderers))

    def __repr__(self):
        return '<%(mod)s.%(cls)s renderers=%(renderers)r>' % {
            'mod': self.__module__,
            'cls': self.__class__.__name__,
            'renderers': self.renderers,
        }

    def is_sitemap(self, cls):
        attrs = ('limit', 'protocol', 'items', 'get_urls')
        return all(hasattr(cls, x) for x in attrs)

    def is_medusa_renderer(self, cls):
        attrs = ('get_paths', 'generate')
        return all(hasattr(cls, x) for x in attrs)

    def is_feed(self, cls):
        attrs = ('get_feed', 'feed_type')
        return all(hasattr(cls, x) for x in attrs)

    def get_renderers(self, renderers=None):
        if renderers is None:
            try:
                renderers = settings.JACKFROST_RENDERERS
            except AttributeError:
                raise ImproperlyConfigured(
                    "You have not set `JACKFROST_RENDERERS` in your "
                    "project's settings")
        for renderer in renderers:
            if isinstance(renderer, six.string_types):
                renderer_cls = import_string(renderer)
            else:
                renderer_cls = renderer

            if self.is_sitemap(cls=renderer_cls):
                renderer_cls = SitemapRenderer(cls=renderer_cls)
            elif self.is_medusa_renderer(cls=renderer_cls):
                renderer_cls = MedusaRenderer(cls=renderer_cls)
            elif self.is_feed(cls=renderer_cls):
                renderer_cls = FeedRenderer(cls=renderer_cls)
            yield renderer_cls

    def get_urls(self):
        urls = set()
        for renderer in self.renderers:
            _cls_or_func_result = renderer()
            # if it was a class, we still need to call __call__
            if callable(_cls_or_func_result):
                _cls_or_func_result = _cls_or_func_result()
            # ensure it's evaluated, incase the renderer is a generator which
            # doesn't wrap itself ...
            for url in _cls_or_func_result:
                current_url = force_text(url)
                if not is_url_usable(url=current_url):
                    raise CollectionError(
                        "Renderer %(renderer)s provided the URL '%(url)s' "
                        "which does not end in a forward-slash ('/'), nor "
                        "does it have a file extension." % {
                            'renderer': renderer, 'url': current_url})
                urls.add(current_url)
        return urls

    def __call__(self):
        return self.get_urls()



def collect(renderers=None):
    return URLCollector(renderers=renderers)()

def read(urls):
    return URLReader(urls=urls)()

def write(data):
    return URLWriter(data=data)()


# Originally: https://gist.github.com/kezabelle/6683315
class ChunkingPaginator(Paginator):
    def chunked_objects(self):
        for page in self.page_range:
            for obj in self.page(page):
                yield obj


class ModelRenderer(object):
    """
    If you just want to render a queryset out, and the model has
    appropriate methods, just subclass this ...
    """
    __slots__ = ()

    def get_model(self):
        raise NotImplementedError("You need to override this ")

    def get_queryset(self):
        return self.get_model().objects.all()

    def get_paginated_queryset(self):
        """
        By default we avoid consuming too much of the database at once, even
        though it means more queries overall.
        You can just replace this with
        return self.get_queryset().iterator() or something if you want.
        """
        return ChunkingPaginator(self.get_queryset(), 50).chunked_objects()

    def _get_urls(self):
        for obj in self.get_paginated_queryset():
            # on the off-chance the app knows it may want to not build things...
            if hasattr(obj, 'jackfrost_can_build'):
                if obj.jackfrost_can_build() is False:
                    continue

            if hasattr(obj, 'jackfrost_urls'):
                for jf_url in obj.jackfrost_urls():
                    yield jf_url
            elif hasattr(obj, 'get_absolute_url'):
                yield obj.get_absolute_url()
            else:
                logger.warning("{obj!r} has no `get_absolute_url` method")
            if hasattr(obj, 'get_list_url'):
                yield obj.get_list_url()

    def get_urls(self):
        return self._get_urls()

    def __call__(self):
        return frozenset(self.get_urls())


class SitemapRenderer(object):
    """
    Given a standard Django Sitemap class, this exposes enough functionality
    to be a renderer of the URLs represented by the Sitemap's location() method.
    """
    __slots__ = ('sitemap_cls',)

    def __init__(self, cls):
        self.sitemap_cls = cls

    def __repr__(self):
        return '<%(mod)s.%(cls)s sitemap_cls=%(medusa)r>' % {
            'mod': self.__module__,
            'cls': self.__class__.__name__,
            'medusa': self.sitemap_cls,
        }

    def get_urls(self):
        sitemap = self.sitemap_cls()
        for page in sitemap.paginator.page_range:
            for result in sitemap.get_urls(page=page):
                if 'location' in result:
                    urlparts = urlparse(result['location'])
                    url = urlparts.path
                    yield url

    def __call__(self):
        return frozenset(self.get_urls())


class MedusaRenderer(object):
    __slots__ = ('medusa_cls',)

    def __init__(self, cls):
        self.medusa_cls = cls

    def __repr__(self):
        return '<%(mod)s.%(cls)s medusa_cls=%(medusa)r>' % {
            'mod': self.__module__,
            'cls': self.__class__.__name__,
            'medusa': self.medusa_cls,
        }

    def get_urls(self):
        return self.medusa_cls().get_paths()

    def __call__(self):
        return frozenset(self.get_urls())


class FeedRenderer(object):
    __slots__ = ('feed_cls',)

    def __init__(self, cls):
        self.feed_cls = cls

    def __repr__(self):
        return '<%(mod)s.%(cls)s feed_cls=%(medusa)r>' % {
            'mod': self.__module__,
            'cls': self.__class__.__name__,
            'medusa': self.feed_cls,
        }

    def get_urls(self):
        feed = self.feed_cls()
        # I have *NO* understanding of why the attribute has been renamed
        # ... what magic is happening here?
        for item in feed._Feed__get_dynamic_attr('items', None):
            yield feed._Feed__get_dynamic_attr('item_link', item)

    def __call__(self):
        return frozenset(self.get_urls())


